File: app/src/main/java/com/cpma/app/main/activities/LoginActivity.java
package com.cpma.app.main.activities;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.cpma.app.main.services.CPMAService;
import com.cpma.app.R;
import com.cpma.app.model.request.DeviceRequest;
import com.cpma.app.model.response.JWTResponse;
import com.cpma.app.model.request.LoginRequest;
import com.cpma.app.model.response.MessageResponse;
import com.cpma.app.repository.DeviceAPI;
import com.cpma.app.utils.EncryptedPreferencesProvider;
import com.cpma.app.utils.ScreenSizeProvider;
import com.cpma.app.utils.retrofit.RetrofitProvider;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LoginActivity extends AppCompatActivity {

    private EditText usernameEditText;
    private EditText passwordEditText;
    private Button loginButton;
    private ProgressBar loadingProgressBar;
    private EncryptedPreferencesProvider encryptedPreferencesProvider;
    private ScreenSizeProvider screenSizeProvider;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Intent intent = new Intent(LoginActivity.this, CPMAService.class);
        stopService(intent);

        setContentView(R.layout.activity_login);
        inicializeFields();

       if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                            ActivityCompat.requestPermissions((Activity) this, new String[]{android.Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION}, 101);
                        }


        loginButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (!validateUsername() | !validatePassword()) {
                    return;
                }else{
                    loadingProgressBar.setVisibility(View.VISIBLE);
                    doLoginRequest();
                }
            }
        });
    }

    private void doLoginRequest() {
        DeviceAPI deviceAPI = RetrofitProvider.createService(DeviceAPI.class);
        Call<JWTResponse> call = deviceAPI.login(new LoginRequest(usernameEditText.getText().toString(), passwordEditText.getText().toString()));
        call.enqueue(new Callback<JWTResponse>() {
            @Override
            public void onResponse(Call<JWTResponse> call, Response<JWTResponse> response) {
                if (response.isSuccessful()) {
                    saveCredentials();
                    JWTResponse jwt = response.body();
                    String token = (jwt.getTokenType() + " "+jwt.getAccessToken());
                    saveDeviceData(RetrofitProvider.createService(DeviceAPI.class, token));
                } else {
                    Gson gson = new Gson();
                    Type type = new TypeToken<MessageResponse>() {}.getType();
                    MessageResponse errorResponse = gson.fromJson(response.errorBody().charStream(),type);
                    Toast.makeText(LoginActivity.this, errorResponse.getMessage(), Toast.LENGTH_SHORT).show();
                    loadingProgressBar.setVisibility(View.GONE);
                }
            }

            @Override
            public void onFailure(Call<JWTResponse> call, Throwable t) {
                Toast.makeText(LoginActivity.this, t.getMessage() , Toast.LENGTH_SHORT).show();
                Log.e("Login method:", t.getMessage());
                loadingProgressBar.setVisibility(View.GONE);
            }
        });
    }

    private void saveDeviceData(DeviceAPI deviceAPI) {
        Call<ResponseBody> call = deviceAPI.saveDevice(createDeviceObject());
        call.enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                if (response.isSuccessful()) {
                    loadingProgressBar.setVisibility(View.GONE);
                    Intent serviceIntent = new Intent(getApplicationContext(), CPMAService.class);
                    ContextCompat.startForegroundService(getApplicationContext(), serviceIntent);
                    finish();
                } else {
                    Toast.makeText(LoginActivity.this, response.message() , Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<ResponseBody> call, Throwable t) {
                Toast.makeText(LoginActivity.this, t.getMessage() , Toast.LENGTH_SHORT).show();
                Log.e("Save device method:", t.getMessage());
            }
        });

    }

    private void saveCredentials() {
        encryptedPreferencesProvider.getSharedPreferences(getApplicationContext()).edit()
        .putString("Username", usernameEditText.getText().toString())
        .putString("Password", passwordEditText.getText().toString())
        .apply();
    }

    private void inicializeFields() {
        usernameEditText = findViewById(R.id.username);
        passwordEditText = findViewById(R.id.password);
        loginButton = findViewById(R.id.login);
        loadingProgressBar = findViewById(R.id.loading);

        SharedPreferences prfs = encryptedPreferencesProvider.getSharedPreferences(getApplicationContext());
        if (prfs != null) {
            usernameEditText.setText(prfs.getString("Username", ""));
            passwordEditText.setText(prfs.getString("Password", ""));
        }
    }

    private boolean validateUsername() {
        String username = usernameEditText.getText().toString().trim();
        if (username.isEmpty()) {
            usernameEditText.setError("Username can't be empty!");
            return false;
        } else {
            usernameEditText.setError(null);
            return true;
        }
    }

    private boolean validatePassword() {
        String password = passwordEditText.getText().toString().trim();
        if (password.isEmpty()) {
            passwordEditText.setError("Password field can't be empty!");
            return false;
        } else if (password.length() < 6) {
            passwordEditText.setError("Password is too short!");
            return false;
        } else {
            passwordEditText.setError(null);
            return true;
        }
    }



    private DeviceRequest createDeviceObject() {
        DeviceRequest dev = new DeviceRequest();

        String android_id = Settings.Secure.getString(getApplicationContext().getContentResolver(), Settings.Secure.ANDROID_ID);
        if(android_id != null && !android_id.isEmpty())
            dev.setAndroidID(android_id);

        String manufacturer = Build.MANUFACTURER;
        if(manufacturer != null && !manufacturer.isEmpty())
            dev.setManufacturer(manufacturer);

        String phoneModel = Build.MODEL;
        if(phoneModel != null && !phoneModel.isEmpty())
            dev.setPhoneModel(phoneModel);

        String brand = Build.BRAND;
        if(brand != null && !brand.isEmpty())
            dev.setBrand(brand);

        String product = Build.PRODUCT;
        if(product != null && !product.isEmpty())
            dev.setProduct(product);

        String serial = Build.SERIAL;
        if(serial != null && !serial.isEmpty())
            dev.setSerial(serial);

        String deviceVersion = String.valueOf(Build.VERSION.SDK_INT);
        if(deviceVersion != null && !deviceVersion.isEmpty())
            dev.setDeviceVersion(deviceVersion);

        int version = Build.VERSION.SDK_INT;
        if(version != 0)
            dev.setVersion(version);

        String versionRelease = Build.VERSION.RELEASE;
        if(versionRelease != null && !versionRelease.isEmpty())
            dev.setVersionRelease(versionRelease);

        int width = screenSizeProvider.getWidth(this);
        if(width != 0)
            dev.setWidth(width);

        int height = screenSizeProvider.getHeight(this);
        if(height != 0)
            dev.setHeight(height);

        return dev;
    }
}

File: app/src/main/java/com/cpma/app/main/activities/LoginActivity.kt
package com.cpma.app.main.activities

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.cpma.app.databinding.ActivityLoginBinding
import com.cpma.app.main.services.CPMAService
import com.cpma.app.model.request.DeviceRequest
import com.cpma.app.model.request.LoginRequest
import com.cpma.app.model.response.JWTResponse
import com.cpma.app.model.response.MessageResponse
import com.cpma.app.repository.DeviceAPI
import com.cpma.app.utils.EncryptedPreferencesProvider
import com.cpma.app.utils.getScreenHeight
import com.cpma.app.utils.getScreenWidth
import com.cpma.app.utils.retrofit.AuthenticationInterceptor
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import dagger.hilt.android.AndroidEntryPoint
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import javax.inject.Inject

@AndroidEntryPoint
class LoginActivity : AppCompatActivity() {

    @Inject
    lateinit var deviceAPI: DeviceAPI

    @Inject
    lateinit var authenticationInterceptor: AuthenticationInterceptor

    @Inject
    lateinit var encryptedPreferencesProvider: EncryptedPreferencesProvider

    private lateinit var binding: ActivityLoginBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val intent = Intent(this@LoginActivity, CPMAService::class.java)
        stopService(intent)

        inicializeFields()
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions((this as Activity), arrayOf(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION), 101)
        }

        binding.login.setOnClickListener(View.OnClickListener {
            if (!validateUsername() or !validatePassword()) {
                return@OnClickListener
            } else {
               binding.loading.visibility = View.VISIBLE
                doLoginRequest()
            }
        })
    }

    private fun doLoginRequest() {
        val call = deviceAPI.login(LoginRequest(binding.username.text.toString(), binding.password.text.toString()))
        call.enqueue(object : Callback<JWTResponse> {
            override fun onResponse(call: Call<JWTResponse>, response: Response<JWTResponse>) {
                if (response.isSuccessful) {
                    saveCredentials()
                    val jwt = response.body()
                    authenticationInterceptor.setAuthToken("${jwt!!.tokenType} ${jwt.accessToken}")
                    saveDeviceData()
                } else {
                    val gson = Gson()
                    val type = object : TypeToken<MessageResponse>() {}.type
                    val (message) = gson.fromJson<MessageResponse>(response.errorBody()!!.charStream(), type)
                    Toast.makeText(this@LoginActivity, message, Toast.LENGTH_SHORT).show()
                    binding.loading.visibility = View.GONE
                }
            }

            override fun onFailure(call: Call<JWTResponse?>, t: Throwable) {
                Toast.makeText(this@LoginActivity, t.message, Toast.LENGTH_SHORT).show()
                Log.e("Login method:", t.message)
                binding.loading.visibility = View.GONE
            }
        })
    }

    private fun saveDeviceData() {
        val call = deviceAPI.saveDevice(createDeviceObject())
        call.enqueue(object : Callback<ResponseBody> {
            override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
                if (response.isSuccessful) {
                    binding.loading.visibility = View.GONE
                    val serviceIntent = Intent(applicationContext, CPMAService::class.java)
                    ContextCompat.startForegroundService(applicationContext, serviceIntent)
                    finish()
                } else {
                    Toast.makeText(this@LoginActivity, response.message(), Toast.LENGTH_SHORT).show()
                }
            }
            override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
                Toast.makeText(this@LoginActivity, t.message, Toast.LENGTH_SHORT).show()
                Log.e("Save device method:", t.message)
            }
        })
    }

    private fun saveCredentials() {
        encryptedPreferencesProvider.saveToEncryptedStorage("Username", binding.username.text.toString())
        encryptedPreferencesProvider.saveToEncryptedStorage("Password", binding.password.text.toString())
    }

    private fun inicializeFields() {
        binding.username.setText(encryptedPreferencesProvider.readFromEncryptedStorage("Username"))
        binding.password.setText(encryptedPreferencesProvider.readFromEncryptedStorage("Password"))
    }

    private fun validateUsername(): Boolean {
        val username = binding.username.text.toString().trim { it <= ' ' }
        return if (username.isEmpty()) {
            binding.username.error = "Username can't be empty!"
            false
        } else {
            binding.username.error = null
            true
        }
    }

    private fun validatePassword(): Boolean {
        val password = binding.password.text.toString().trim { it <= ' ' }
        return if (password.isEmpty()) {
            binding.password.error = "Password field can't be empty!"
            false
        } else if (password.length < 6) {
            binding.password.error = "Password is too short!"
            false
        } else {
            binding.password.error = null
            true
        }
    }

    private fun createDeviceObject(): DeviceRequest {
        val dev = DeviceRequest()
        val android_id = Settings.Secure.getString(applicationContext.contentResolver, Settings.Secure.ANDROID_ID)
        if (android_id != null && !android_id.isEmpty()) dev.androidID = android_id
        val manufacturer = Build.MANUFACTURER
        if (manufacturer != null && !manufacturer.isEmpty()) dev.manufacturer = manufacturer
        val phoneModel = Build.MODEL
        if (phoneModel != null && !phoneModel.isEmpty()) dev.phoneModel = phoneModel
        val brand = Build.BRAND
        if (brand != null && !brand.isEmpty()) dev.brand = brand
        val product = Build.PRODUCT
        if (product != null && !product.isEmpty()) dev.product = product
        dev.deviceVersion = Build.VERSION.SDK_INT.toString()
        val version = Build.VERSION.SDK_INT
        if (version != 0) dev.version = version
        val versionRelease = Build.VERSION.RELEASE
        if (versionRelease != null && !versionRelease.isEmpty()) dev.versionRelease = versionRelease
        val width = getScreenWidth(this)
        if (width != 0) dev.width = width
        val height = getScreenHeight(this)
        if (height != 0) dev.height = height
        return dev
    }
}